### PR TITLE
[SPARK-29470][BUILD] Update plugins to latest versions

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR updates plugins to latest versions.

### Why are the changes needed?

This brings bug fixes like the following.
- https://issues.apache.org/jira/projects/MCOMPILER/versions/12343484 (maven-compiler-plugin)
- https://issues.apache.org/jira/projects/MJAVADOC/versions/12345060 (maven-javadoc-plugin)
- https://issues.apache.org/jira/projects/MCHECKSTYLE/versions/12342397 (maven-checkstyle-plugin)
- https://checkstyle.sourceforge.io/releasenotes.html#Release_8.25 (checkstyle)
- https://checkstyle.sourceforge.io/releasenotes.html#Release_8.24 (checkstyle)

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Pass the Jenkins building and testing with the existing code.

### CODE REVIEW

File: dev/checkstyle.xml [hunk 1]
```diff
@@ -91,10 +96,6 @@
             <property name="allowByTailComment" value="true"/>
             <property name="allowNonPrintableEscapes" value="true"/>
         </module>
-        <module name="LineLength">
-            <property name="max" value="100"/>
-            <property name="ignorePattern" value="^package.*|^import.*|a href|href|http://|https://|ftp://"/>
-        </module>
```
Comment (dongjoon-hyun) [inline dev/checkstyle.xml]: This is banned at `8.24`.
- https://github.com/checkstyle/checkstyle/issues/2116

File: core/src/main/resources/org/apache/spark/log4j-defaults.properties [hunk 2]
```diff
@@ -33,7 +33,8 @@ log4j.logger.org.sparkproject.jetty.util.component.AbstractLifeCycle=ERROR
 log4j.logger.org.apache.spark.repl.SparkIMain$exprTyper=INFO
 log4j.logger.org.apache.spark.repl.SparkILoop$SparkILoopInterpreter=INFO
 
-# SPARK-9183: Settings to avoid annoying messages when looking up nonexistent UDFs in SparkSQL with Hive support
+# SPARK-9183: Settings to avoid annoying messages when looking up nonexistent UDFs
+# in SparkSQL with Hive support
```
Comment (dongjoon-hyun) [inline core/src/main/resources/org/apache/spark/log4j-defaults.properties]:
```
$ dev/lint-java
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=384m; support was removed in 8.0
Using `mvn` from path: /usr/local/bin/mvn
Checkstyle checks failed at following occurrences:
[ERROR] src/main/resources/org/apache/spark/log4j-defaults.properties:[36] (sizes) LineLength: Line is longer than 100 characters (found 112).
```

**Comment by dongjoon-hyun** [comment]
Hi, @viirya . Could you review this, please?

File: project/plugins.sbt [hunk 3]
```diff
@@ -18,7 +18,7 @@
 addSbtPlugin("com.etsy" % "sbt-checkstyle-plugin" % "3.1.1")
 
 // sbt-checkstyle-plugin uses an old version of checkstyle. Match it to Maven's.
-libraryDependencies += "com.puppycrawl.tools" % "checkstyle" % "8.23"
+libraryDependencies += "com.puppycrawl.tools" % "checkstyle" % "8.25"
```
Comment (viirya) [inline project/plugins.sbt]: Why it is an old version? Looks like it is very new version?

https://mvnrepository.com/artifact/com.puppycrawl.tools/checkstyle

**Review by viirya — APPROVED** [review]

File: project/plugins.sbt (hunk 3 above)
Comment (dongjoon-hyun) [reply project/plugins.sbt]: The description `sbt-checkstyle-plugin uses an old version` means that `sbt-checkstyle-plugin` depends on the old version and we need to override it here. :)

**Comment by dongjoon-hyun** [comment]
Thank you so much for review and approval, @viirya !
Merged to master.

File: project/plugins.sbt (hunk 3 above)
Comment (viirya) [reply project/plugins.sbt]: I see. Thanks for clarifying.

**Pull request closed without merging**